### PR TITLE
Added posibility of configure the input value for the RegExFilter.

### PR DESCRIPTION
**Pull request by estebanwasinger**
* Added 'value' optional attribute for the regexfilter
* Added Integration tests for regexfilter

### CODE REVIEW

**Comment by afelisatti** [comment]
Please add the Jira issue id to the commit message.

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 1]
```diff
@@ -19,17 +23,19 @@
...
  */
-public class RegExFilter implements Filter, ObjectFilter
+public class RegExFilter implements Filter, ObjectFilter, MuleContextAware
 {
+
     private static final int NO_FLAGS = 0;
     protected transient Log logger = LogFactory.getLog(getClass());
 
     private Pattern pattern;
+    private String value;
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: Check out org.mule.util.AttributeEvaluator

File: core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java [hunk 2]
```diff
@@ -139,12 +153,47 @@ public void testEqualsWithDifferentPattern()
     @Test
     public void testEqualsWithEqualPatternAndDifferentFlags()
     {
-        RegExFilter filter1 = new RegExFilter(PATTERN, Pattern.DOTALL);
-        RegExFilter filter2 = new RegExFilter(PATTERN, Pattern.CASE_INSENSITIVE);
+        RegExFilter filter1 = new RegExFilter(PATTERN, null, Pattern.DOTALL);
```
Comment (afelisatti) [inline core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java]: You can avoid this with another constructor.

File: tests/integration/src/test/resources/org/mule/test/integration/filter/regex-test.xml [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<mule xmlns="http://www.mulesoft.org/schema/mule/core"
+		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		xsi:schemaLocation="
+							http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+							">
+
+	<flow name="regex-filter-default-payload-positive">
+        <set-payload value="#['run with the mules']"/>
```
Comment (afelisatti) [inline tests/integration/src/test/resources/org/mule/test/integration/filter/regex-test.xml]: You can avoid the almost duplicate flows setting the payload in the tests. That will also make the tests clearer.

File: tests/integration/src/test/java/org/mule/test/integration/filter/RegexFilterIntegrationTestCase.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+{
+
+    @Override
+    protected String getConfigFile()
+    {
+        return "org/mule/test/integration/filter/regex-test.xml";
+    }
+
+    @Test
+    public void testRegexFilterDefaultPayloadPositive() throws Exception
```
Comment (afelisatti) [inline tests/integration/src/test/java/org/mule/test/integration/filter/RegexFilterIntegrationTestCase.java]: Please add clearer names for the tests. Avoid the "test" and positive/negative.

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 5]
```diff
@@ -117,11 +137,27 @@ public void setFlags(int flags)
         this.pattern = (this.pattern != null ? Pattern.compile(pattern.pattern(), flags) : null);
     }
 
+    public String getValue()
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: Why do you have this getter?

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 6]
```diff
@@ -117,11 +137,27 @@ public void setFlags(int flags)
...
+        return value;
+    }
+
+    public void setValue(String value)
+    {
+        this.value = value;
+    }
+
     @Override
     public boolean equals(Object obj)
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: You should consider the value here now (hash as well).

File: core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java [hunk 7]
```diff
@@ -139,12 +153,47 @@ public void testEqualsWithDifferentPattern()
...
-        filter1 = new RegExFilter(PATTERN, Pattern.DOTALL);
-        filter2 = new RegExFilter(PATTERN, Pattern.DOTALL);
+        filter1 = new RegExFilter(PATTERN, null, Pattern.DOTALL);
+        filter2 = new RegExFilter(PATTERN, null, Pattern.DOTALL);
         assertEquals(filter1, filter2);
     }
+
+
+    @Test
+    public void testRegexValueFromMelPayloadPositive()
```
Comment (afelisatti) [inline core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java]: Same as the other test names.

File: core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java [hunk 8]
```diff
@@ -6,28 +6,42 @@
...
+public class RegExFilterTestCase extends AbstractMuleContextTestCase
 {
+
     private static final String PATTERN = "(.*) brown fox";
+    private RegExFilter regExWithValue;
+
+    @Before
+    public void setUp() throws Exception
+    {
+        super.setUpMuleContext();
```
Comment (afelisatti) [inline core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java]: Doesn't this happened already in AbstractMuleContextTestCase?

**Comment by estebanwasinger** [comment]
Updated commit fixing all the issues:
* Value attribute now it is a AttributeEvaluator
* Test conventions
* Update equals() and hashCode() to use the RawValue of the AttributeEvaluator

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 9]
```diff
@@ -117,21 +144,52 @@ public void setFlags(int flags)
...
     @Override
     public int hashCode()
     {
-        return hash(new Object[]{this.getClass(), pattern});
+        return hash(new Object[] {this.getClass(), pattern, value.getRawValue()});
+    }
+
+    private boolean attributeWrapperRawValueEqualityValue(AttributeEvaluator value1, AttributeEvaluator value2)
+    {
+        if (value1.getRawValue() == null && value2.getRawValue() == null)
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: Isn't this the same as if (value1.getRawValue() == null) return value2.getRawValue() == null?

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 10]
```diff
@@ -117,21 +144,52 @@ public void setFlags(int flags)
...
+        if (obj == null || getClass() != obj.getClass())
+        {
+            return false;
+        }
 
         final RegExFilter other = (RegExFilter) obj;
         boolean patternsAreEqual = ClassUtils.equal(pattern.pattern(), other.pattern.pattern());
         boolean flagsAreEqual = (flags == other.flags);
-        return (patternsAreEqual && flagsAreEqual);
+        boolean equalValues = attributeWrapperRawValueEqualityValue(value, other.value);
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: valuesAreEqual?

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 11]
```diff
@@ -117,21 +144,52 @@ public void setFlags(int flags)
...
     }
 
     @Override
     public int hashCode()
     {
-        return hash(new Object[]{this.getClass(), pattern});
+        return hash(new Object[] {this.getClass(), pattern, value.getRawValue()});
+    }
+
+    private boolean attributeWrapperRawValueEqualityValue(AttributeEvaluator value1, AttributeEvaluator value2)
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: This name is very confusing.

File: core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java [hunk 12]
```diff
@@ -35,15 +44,15 @@ public void testRegexFilterNoPattern()
         // start with default
         RegExFilter filter = new RegExFilter();
         assertNull(filter.getPattern());
-        assertFalse(filter.accept("No tengo dinero"));
+        assertThat(filter.accept("No tengo dinero"), is(Boolean.FALSE));
```
Comment (afelisatti) [inline core/src/test/java/org/mule/routing/filters/RegExFilterTestCase.java]: Primitives.

File: tests/integration/src/test/java/org/mule/test/integration/filter/RegexFilterIntegrationTestCase.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Before;
+import org.junit.Test;
+
+public class RegexFilterIntegrationTestCase extends FunctionalTestCase
+{
+
+    Map<String, String> map;
```
Comment (afelisatti) [inline tests/integration/src/test/java/org/mule/test/integration/filter/RegexFilterIntegrationTestCase.java]: inboundProperties?

File: tests/integration/src/test/resources/org/mule/test/integration/filter/regex-test.xml [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xsi:schemaLocation="
+            http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
+      ">
+
+    <flow name="matches-using-default-value">
+        <regex-filter pattern="(\w)* with the mules"/>
+    </flow>
+
+    <flow name="not-matches-using-default-value">
```
Comment (afelisatti) [inline tests/integration/src/test/resources/org/mule/test/integration/filter/regex-test.xml]: This flow is redundant.

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 15]
```diff
@@ -117,21 +144,45 @@ public void setFlags(int flags)
...
     public int hashCode()
     {
-        return hash(new Object[]{this.getClass(), pattern});
+        return hash(new Object[] {this.getClass(), pattern, value.getRawValue()});
+    }
+
+    private boolean areRawValueEquals(AttributeEvaluator value1, AttributeEvaluator value2)
+    {
+        if (value1.getRawValue() == null) return value2.getRawValue() == null;
+        return value2.getRawValue() != null && value1.getRawValue().equals(value2.getRawValue());
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: No need to check for null.

File: core/src/main/java/org/mule/routing/filters/RegExFilter.java [hunk 16]
```diff
@@ -117,21 +144,45 @@ public void setFlags(int flags)
...
     @Override
     public int hashCode()
     {
-        return hash(new Object[]{this.getClass(), pattern});
+        return hash(new Object[] {this.getClass(), pattern, value.getRawValue()});
+    }
+
+    private boolean areRawValueEquals(AttributeEvaluator value1, AttributeEvaluator value2)
+    {
+        if (value1.getRawValue() == null) return value2.getRawValue() == null;
```
Comment (afelisatti) [inline core/src/main/java/org/mule/routing/filters/RegExFilter.java]: Please add braces and proper formatting. Sorry for the misleading comment.

**Comment by afelisatti** [comment]
Thank you @estebanwasinger for your contribution!

**Pull request merged**